exporter/opencensusexporter/opencensus.go

```diff
@@ -16,183 +16,241 @@ package opencensusexporter
-	"go.opentelemetry.io/collector/config/configmodels"
 	"go.opentelemetry.io/collector/consumer/consumerdata"
 	"go.opentelemetry.io/collector/exporter/exporterhelper"
 )
 
-type ocAgentExporter struct {
-	exporters chan *ocagent.Exporter
+// See https://godoc.org/google.golang.org/grpc#ClientConn.NewStream
+// why we need to keep the cancel func to cancel the stream
+type tracesClientWithCancel struct {
+	cancel context.CancelFunc
+	tsec   agenttracepb.TraceService_ExportClient
 }
 
-type ocExporterErrorCode int
-type ocExporterError struct {
-	code ocExporterErrorCode
-	msg  string
+// See https://godoc.org/google.golang.org/grpc#ClientConn.NewStream
+// why we need to keep the cancel func to cancel the stream
+type metricsClientWithCancel struct {
+	cancel context.CancelFunc
+	msec   agentmetricspb.MetricsService_ExportClient
 }
 
-var _ error = (*ocExporterError)(nil)
-
-func (e *ocExporterError) Error() string {
-	return e.msg
+type ocExporter struct {
+	cfg *Config
+	// gRPC clients and connection.
+	traceSvcClient   agenttracepb.TraceServiceClient
+	metricsSvcClient agentmetricspb.MetricsServiceClient
+	tracesClients    chan *tracesClientWithCancel
+	metricsClients   chan *metricsClientWithCancel
+	grpcClientConn   *grpc.ClientConn
+	metadata         metadata.MD
+	waitForReady     bool
 }
 
-const (
-	defaultNumWorkers int = 2
-
-	_ ocExporterErrorCode = iota // skip 0
-	// errEndpointRequired indicates that this exporter was not provided with an endpoint in its config.
-	errEndpointRequired
-	// errUnsupportedCompressionType indicates that this exporter was provided with a compression protocol it does not support.
-	errUnsupportedCompressionType
-	// errUnableToGetTLSCreds indicates that this exporter could not read the provided TLS credentials.
-	errUnableToGetTLSCreds
-	// errAlreadyStopped indicates that the exporter was already stopped.
-	errAlreadyStopped
-)
+func newOcExporter(ctx context.Context, cfg *Config) (*ocExporter, error) {
+	if cfg.Endpoint == "" {
+		return nil, errors.New("OpenCensus exporter cfg requires an Endpoint")
+	}
+
+	if cfg.NumWorkers <= 0 {
+		return nil, errors.New("OpenCensus exporter cfg requires at least one worker")
+	}
 
-// NewTraceExporter creates an Open Census trace exporter.
-func NewTraceExporter(logger *zap.Logger, config configmodels.Exporter, opts ...ocagent.ExporterOption) (component.TraceExporterOld, error) {
-	oce, err := createOCAgentExporter(logger, config, opts...)
+	dialOpts, err := cfg.GRPCClientSettings.ToDialOptions()
 	if err != nil {
 		return nil, err
 	}
-	oexp, err := exporterhelper.NewTraceExporterOld(
-		config,
-		oce.PushTraceData,
-		exporterhelper.WithShutdown(oce.Shutdown))
-	if err != nil {
+
+	// Cannot use WaitForReady in the calls because we don't have a "deadline"
+	// set for every message, tradeoff to use streaming, also does not help with
+	// individual messages. So to achieve something similar, we block to ensure we connect.
+	if cfg.WaitForReady {
```

Comment:
So this means if WaitForReady=true and we cannot connect then startup will be aborted, right? Is this an expected behavior?

Reply:
Removed for the moment. Will think about this and add later, was mostly useful for some tests, but don't have to support it now.